Build failure: rustdesk{,-flutter}

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

1. `nix build nixpkgs/0d534853a55b5d02a4ababa1d71921ce8f0aee4c#rustdesk-flutter`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/292246311#tabs-summary

### Relevant log output

```console
error[E0583]: file not found for module `internal`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:16:1
   |
16 | mod internal;
   | ^^^^^^^^^^^^^
   |
   = help: to create the module `internal`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/internal.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/internal/mod.rs"
   = note: if there is a `mod internal` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xlib`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:19:1
   |
19 | pub mod xlib;
   | ^^^^^^^^^^^^^
   |
   = help: to create the module `xlib`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xlib.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xlib/mod.rs"
   = note: if there is a `mod xlib` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `dpms`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:21:1
   |
21 | pub mod dpms;
   | ^^^^^^^^^^^^^
   |
   = help: to create the module `dpms`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/dpms.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/dpms/mod.rs"
   = note: if there is a `mod dpms` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `glx`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:22:1
   |
22 | pub mod glx;
   | ^^^^^^^^^^^^
   |
   = help: to create the module `glx`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/glx.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/glx/mod.rs"
   = note: if there is a `mod glx` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `keysym`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:23:1
   |
23 | pub mod keysym;
   | ^^^^^^^^^^^^^^^
   |
   = help: to create the module `keysym`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/keysym.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/keysym/mod.rs"
   = note: if there is a `mod keysym` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xcursor`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:24:1
   |
24 | pub mod xcursor;
   | ^^^^^^^^^^^^^^^^
   |
   = help: to create the module `xcursor`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xcursor.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xcursor/mod.rs"
   = note: if there is a `mod xcursor` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xf86vmode`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:25:1
   |
25 | pub mod xf86vmode;
   | ^^^^^^^^^^^^^^^^^^
   |
   = help: to create the module `xf86vmode`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xf86vmode.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xf86vmode/mod.rs"
   = note: if there is a `mod xf86vmode` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xfixes`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:26:1
   |
26 | pub mod xfixes;
   | ^^^^^^^^^^^^^^^
   |
   = help: to create the module `xfixes`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xfixes.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xfixes/mod.rs"
   = note: if there is a `mod xfixes` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xft`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:27:1
   |
27 | pub mod xft;
   | ^^^^^^^^^^^^
   |
   = help: to create the module `xft`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xft.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xft/mod.rs"
   = note: if there is a `mod xft` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xinerama`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:28:1
   |
28 | pub mod xinerama;
   | ^^^^^^^^^^^^^^^^^
   |
   = help: to create the module `xinerama`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xinerama.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xinerama/mod.rs"
   = note: if there is a `mod xinerama` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xinput`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:29:1
   |
29 | pub mod xinput;
   | ^^^^^^^^^^^^^^^
   |
   = help: to create the module `xinput`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xinput.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xinput/mod.rs"
   = note: if there is a `mod xinput` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xinput2`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:30:1
   |
30 | pub mod xinput2;
   | ^^^^^^^^^^^^^^^^
   |
   = help: to create the module `xinput2`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xinput2.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xinput2/mod.rs"
   = note: if there is a `mod xinput2` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xlib_xcb`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:31:1
   |
31 | pub mod xlib_xcb;
   | ^^^^^^^^^^^^^^^^^
   |
   = help: to create the module `xlib_xcb`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xlib_xcb.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xlib_xcb/mod.rs"
   = note: if there is a `mod xlib_xcb` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xmd`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:32:1
   |
32 | pub mod xmd;
   | ^^^^^^^^^^^^
   |
   = help: to create the module `xmd`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xmd.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xmd/mod.rs"
   = note: if there is a `mod xmd` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xmu`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:33:1
   |
33 | pub mod xmu;
   | ^^^^^^^^^^^^
   |
   = help: to create the module `xmu`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xmu.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xmu/mod.rs"
   = note: if there is a `mod xmu` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xrandr`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:34:1
   |
34 | pub mod xrandr;
   | ^^^^^^^^^^^^^^^
   |
   = help: to create the module `xrandr`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xrandr.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xrandr/mod.rs"
   = note: if there is a `mod xrandr` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xrecord`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:35:1
   |
35 | pub mod xrecord;
   | ^^^^^^^^^^^^^^^^
   |
   = help: to create the module `xrecord`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xrecord.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xrecord/mod.rs"
   = note: if there is a `mod xrecord` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xrender`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:36:1
   |
36 | pub mod xrender;
   | ^^^^^^^^^^^^^^^^
   |
   = help: to create the module `xrender`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xrender.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xrender/mod.rs"
   = note: if there is a `mod xrender` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xss`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:37:1
   |
37 | pub mod xss;
   | ^^^^^^^^^^^^
   |
   = help: to create the module `xss`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xss.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xss/mod.rs"
   = note: if there is a `mod xss` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xt`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:38:1
   |
38 | pub mod xt;
   | ^^^^^^^^^^^
   |
   = help: to create the module `xt`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xt.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xt/mod.rs"
   = note: if there is a `mod xt` elsewhere in the crate already, import it with `use crate::...` instead

error[E0583]: file not found for module `xtest`
  --> /build/rustdesk-1.3.8-vendor/x11-2.19.0/src/lib.rs:39:1
   |
39 | pub mod xtest;
   | ^^^^^^^^^^^^^^
   |
   = help: to create the module `xtest`, create file "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xtest.rs" or "/build/rustdesk-1.3.8-vendor/x11-2.19.0/src/xtest/mod.rs"
   = note: if there is a `mod xtest` elsewhere in the crate already, import it with `use crate::...` instead

For more information about this error, try `rustc --explain E0583`.
error: could not compile `x11` (lib) due to 21 previous errors
warning: build failed, waiting for other jobs to finish...
```

### Additional context

I'm suspecting that #383902 broke it

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.13.5-zen1, NixOS, 25.05 (Warbler), 25.05.20250309.e3e32b6`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.91.1
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/gep/.config/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/gep/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/gep/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/gep/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/zm68nasdmai8qzgzlb55m10mrl93dzqd-lix-2.91.1/share`
 - nixpkgs: `/nix/store/xq5rfjj1z2r8yx338arajg5vwsxh1fri-source`

### Notify maintainers

@dasJ @Conni2461 @helsinki-Jo 
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/build-support/rust/fetch-cargo-vendor-util.py
+++ b/pkgs/build-support/rust/fetch-cargo-vendor-util.py
@@ -7,6 +7,7 @@ import shutil
 import subprocess
 import sys
 import tomllib
+from os.path import islink, realpath
 from pathlib import Path
 from typing import Any, TypedDict, cast
 from urllib.parse import unquote
@@ -194,11 +195,41 @@ def find_crate_manifest_in_tree(tree: Path, crate_name: str) -> Path:
 
 
 def copy_and_patch_git_crate_subtree(git_tree: Path, crate_name: str, crate_out_dir: Path) -> None:
+
+    # This function will get called by copytree to decide which entries of a directory should be copied
+    # We'll copy everything except symlinks that are invalid
+    def ignore_func(dir_str: str, path_strs: list[str]) -> list[str]:
+        ignorelist: list[str] = []
+
+        dir = Path(realpath(dir_str, strict=True))
+
+        for path_str in path_strs:
+            path = dir / path_str
+            if not islink(path):
+                continue
+
+            # Filter out cyclic symlinks and symlinks pointing at nonexistant files
+            try:
+                target_path = Path(realpath(path, strict=True))
+            except OSError:
+                ignorelist.append(path_str)
+                eprint(f"Failed to resolve symlink, ignoring: {path}")
+                continue
+
+            # Filter out symlinks that point outside of the current crate's base git tree
+            # This can be useful if the nix build sandbox is turned off and there is a symlink to a common absolute path
+            if not target_path.is_relative_to(git_tree):
+                ignorelist.append(path_str)
+                eprint(f"Symlink points outside of the crate's base git tree, ignoring: {path} -> {target_path}")
+                continue
+
+        return ignorelist
+
     crate_manifest_path = find_crate_manifest_in_tree(git_tree, crate_name)
     crate_tree = crate_manifest_path.parent
 
     eprint(f"Copying to {crate_out_dir}")
-    shutil.copytree(crate_tree, crate_out_dir, ignore_dangling_symlinks=True)
+    shutil.copytree(crate_tree, crate_out_dir, ignore=ignore_func)
     crate_out_dir.chmod(0o755)
 
     with open(crate_manifest_path, "r") as f:

--- a/pkgs/build-support/rust/import-cargo-lock.nix
+++ b/pkgs/build-support/rust/import-cargo-lock.nix
@@ -195,7 +195,7 @@ let
         echo Found crate ${pkg.name} at $crateCargoTOML
         tree=$(dirname $crateCargoTOML)
 
-        ${python3Packages.python.interpreter} -c 'import sys, shutil; shutil.copytree(sys.argv[1], sys.argv[2], ignore_dangling_symlinks=True)' "$tree" "$out"
+        cp -prvL "$tree" "$out" || echo "Warning: certain files couldn't be copied!" >&2
         chmod u+w $out
 
         if grep -q workspace "$out/Cargo.toml"; then

